File: src/main/java/ch/grassl/day03/impl/model/Team.java
package ch.grassl.day03.impl.model;

import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;

public class Team {

    private final List<Rucksack> rucksacks;

    public Team(List<Rucksack> rucksacks) {
        this.rucksacks = rucksacks;
    }

    /**
     * Finds the teams badge by removing the duplicates from all items of a rucksack. These items are combined and counted.
     *
     * @return The item that appears in every rucksack
     */
    public Item findBadge() {
        List<Item> teamItems = getCombinedTeamItems();
        Map<Item, Long> itemsCount = getItemsCount(teamItems);

        return itemsCount.entrySet().stream()
                .filter(e -> e.getValue() == rucksacks.size())
                .findFirst()
                .orElseThrow(() -> new IllegalStateException("Found no Badge for Team, but a Team has to have a Badge."))
                .getKey();
    }

    private List<Item> getCombinedTeamItems() {
        return rucksacks.stream()
                .map(Rucksack::getInventory)
                .map(HashSet::new)
                .flatMap(Set::stream)
                .toList();
    }

    private static Map<Item, Long> getItemsCount(List<Item> teamItems) {
        return teamItems.stream()
                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
    }
}

File: src/main/java/ch/grassl/day03/impl/model/Team.kt
package ch.grassl.day03.impl.model

class Team(private val rucksacks: List<Rucksack>) {

    /**
     * Finds the teams badge by removing the duplicates from all items of a rucksack. These items are combined and counted.
     *
     * @return The item that appears in every rucksack
     */
    fun findBadge(): Item {
        val teamItems: List<Item> = getCombinedTeamItems()
        val itemsCount: Map<Item, Int> = getItemsCount(teamItems)

        return itemsCount.entries
            .firstOrNull { it.value == rucksacks.size }
            .let {
                checkNotNull(it)
                it.key
            }
    }

    private fun getCombinedTeamItems(): List<Item> {
        return rucksacks
            .map { it.getInventory() }
            .map { HashSet(it) }
            .flatMap { it.toList() }
    }

    private fun getItemsCount(teamItems: List<Item>): Map<Item, Int> {
        return teamItems
            .groupingBy { it }
            .eachCount()
    }
}